### PR TITLE
Nonlinear solver: deal with failure

### PR DESCRIPTION
**Pull request by tjhei**
This introduces a new setting to decide what should happen if the
nonlinear solver fails to converge.

The old behavior was to "just continue", which is an unsafe choice.

### CODE REVIEW

**Comment by tjhei** [comment]
@jdannberg something like this?

File: source/simulator/core.cc [hunk 1]
```diff
@@ -1767,15 +1767,53 @@ namespace aspect
...
-        // Only solve if we are not in pre-refinement, or we do not want to skip
-        // solving in pre-refinement.
+        // During pre-refinement, do not solve if we are asked to skip it:
         if (! (parameters.skip_solvers_on_initial_refinement
                && pre_refinement_step < parameters.initial_adaptive_refinement))
           {
             start_timestep ();
 
             // then do the core work: assemble systems and solve
-            solve_timestep ();
```
Comment (gassmoeller) [inline source/simulator/core.cc]: I somewhat dislike adding so many lines in the main run loop that are unrelated to it. Conceptionally it is the responsibility of the solve_timestep function to handle non-convergence. Could we move the whole try-catch block into solve_timestep, and modify the code in `run` to the following:
```
const bool retry_timestep = solve_timestep();
if (retry_timestep)
  continue;
```

All of the exception handling can stay the same, but it keep this loop readable.

File: source/simulator/parameters.cc [hunk 2]
```diff
@@ -221,6 +221,14 @@ namespace aspect
                        "The `Newton Stokes' scheme is deprecated and only allowed for reasons of "
                        "backwards compatibility. It is the same as `iterated Advection and Newton Stokes'.");
 
+    prm.declare_entry ("Nonlinear solver failure strategy", "abort program",
+                       Patterns::Selection("continue with next timestep|cut timestep size|abort program"),
+                       "Select the strategy on what to do if the nonlinear solver scheme fails to "
+                       "converge. The options are:\n"
+                       "`continue with next timestep`: ignore error and continue to the next timestep\n"
+                       "`cut timestep size`: reduce the current timestep size and redo the timestep\n"
```
Comment (gassmoeller) [inline source/simulator/parameters.cc]: maybe `by a factor of two`?

File: source/simulator/core.cc [hunk 3]
```diff
@@ -1767,15 +1767,53 @@ namespace aspect
...
+            try
+              {
+                solve_timestep ();
+              }
+            catch (...)
+              {
+                pcout << "ERROR: the solver in the current timestep failed to converge." << std::endl;
```
Comment (gassmoeller) [inline source/simulator/core.cc]: What about
```
pcout << "WARNING: The nonlinear solver in the current timestep failed to converge." << std::endl
          << "Acting according to the parameter 'Nonlinear solver failure strategy'." << std::endl;
```

File: source/simulator/core.cc [hunk 4]
```diff
@@ -1767,15 +1767,53 @@ namespace aspect
...
+                    == Parameters<dim>::NonlinearSolverFailureStrategy::continue_with_next_timestep)
+                  {
+                    // do nothing and continue
+                  }
+                else if (parameters.nonlinear_solver_failure_strategy
+                         == Parameters<dim>::NonlinearSolverFailureStrategy::cut_timestep_size)
+                  {
+                    if (timestep_number == 0)
+                      {
+                        pcout << "Note: we can not cut the timestep in step 0, so we are aborting."
```
Comment (gassmoeller) [inline source/simulator/core.cc]: I would call this an `ERROR:...`

File: include/aspect/simulator.h [hunk 5]
```diff
@@ -123,6 +123,14 @@ namespace aspect
     template <int dim>      class Manager;
   }
 
+  /**
+   * Exception to be thrown when the nonlinear solver needs to many iterations to converge.
```
Comment (gassmoeller) [inline include/aspect/simulator.h]: to -> too

File: source/simulator/core.cc [hunk 6]
```diff
@@ -1767,15 +1767,53 @@ namespace aspect
...
+                              << std::endl;
+                        throw;
+                      }
+
+                    // reduce timestep size and update time to "go back":
+                    const double new_time_step = 0.5*time_step;
+                    pcout << "\tReducing timestep to " << new_time_step << '\n' << std::endl;
```
Comment (jdannberg) [inline source/simulator/core.cc:1802]: This output should be in years or seconds, depending on the parameter.

File: source/simulator/core.cc [hunk 7]
```diff
@@ -1767,15 +1767,53 @@ namespace aspect
...
+
+                    time -= time_step - new_time_step;
+                    time_step = new_time_step;
```
Comment (jdannberg) [inline source/simulator/core.cc:1805]: There's one thing that I think doesn't work well with the current implementation: The time step size is reduced for the current time step, until the solver converges, but the next time step bases the time step size on the CFL number (rather than on what time step size was necessary to actually get the nonlinear solver to converge). So if the problem stays as hard to solve, we need to go though a number of failing nonlinear solver loops for every time step, rather than continuing with the new (smaller) time step that is adapted to the nonlinear problem.

File: tests/nonlinear_failure_strategy_cut.cc [hunk 8]
```diff
@@ -0,0 +1 @@
+#include "../benchmarks/nonlinear_channel_flow/simple_nonlinear.cc"
```
Comment (jdannberg) [inline tests/nonlinear_failure_strategy_cut.cc:1]: It looks like this test problem is some sort of benchmark. So would it make sense to do some postprocessing, so that we see that our new method actually still gives us the correct solution?

File: source/simulator/core.cc (hunk 7 above)
Comment (tjhei) [reply source/simulator/core.cc:1805]: That makes sense, but how do you want to decide when to make bigger timestep again? I don't think there is a good answer here. One thing that might help in an extreme case is not increasing it by more than a factor of say 2 in each step. But in general, this seems to be something that would require more user input.

**Comment by tjhei** [comment]
I am resurrecting this old PR because @danieldouglas92 is interested in it. I rebased and I am curious to see if things still work, but I know I have to make some more changes.

**Comment by tjhei** [comment]
This has been updated to the current main branch and now uses the TimeStepping infrastructure. The test added here is working correctly.

I would be grateful if somebody can take a look at the current state.

**Comment by tjhei** [comment]
@jdannberg ? @gassmoeller ?

**Comment by gassmoeller** [comment]
I will take a look asap.

**Comment by tjhei** [comment]
it helps to include all files. :-)

File: include/aspect/simulator.h (hunk 5 above)
Comment (gassmoeller) [reply include/aspect/simulator.h:172]: too

File: include/aspect/time_stepping/interface.h [hunk 9]
```diff
@@ -234,6 +234,36 @@ namespace aspect
         void
         parse_parameters (ParameterHandler &prm);
 
+        /**
+         * Go through the list of all mesh refinement strategies that have been selected
```
Comment (gassmoeller) [inline include/aspect/time_stepping/interface.h:238]:
```suggestion
         * Go through the list of all time stepping plugins that have been selected
```

File: include/aspect/time_stepping/interface.h [hunk 10]
```diff
@@ -234,6 +234,36 @@ namespace aspect
...
+         * This function can only be called if the given template type (the first template
+         * argument) is a class derived from the Interface class in this namespace.
+         */
+        template <typename PluginType,
+                  typename = typename std::enable_if_t<std::is_base_of<Interface<dim>,PluginType>::value>>
+        bool
+        has_matching_plugin () const;
+
+        /**
+         * Go through the list of all mesh refinement strategies that have been selected
```
Comment (gassmoeller) [inline include/aspect/time_stepping/interface.h:252]:
```suggestion
         * Go through the list of all time stepping plugins that have been selected
```

File: include/aspect/time_stepping/interface.h [hunk 11]
```diff
@@ -234,6 +234,36 @@ namespace aspect
...
+         * in the input file (and are consequently currently active) and see
+         * if one of them has the type specified by the template
+         * argument or can be casted to that type. If so, return a reference
```
Comment (gassmoeller) [inline include/aspect/time_stepping/interface.h:255]:
```suggestion
         * argument or can be cast to that type. If so, return a reference
```

File: include/aspect/time_stepping/interface.h [hunk 12]
```diff
@@ -308,6 +337,50 @@ namespace aspect
...
+
+    template <int dim>
+    template <typename PluginType, typename>
+    inline
+    bool
+    Manager<dim>::has_matching_plugin () const
+    {
+      for (typename std::list<std::unique_ptr<Interface<dim>>>::const_iterator
+           p = active_plugins.begin();
+           p != active_plugins.end(); ++p)
```
Comment (gassmoeller) [inline include/aspect/time_stepping/interface.h:350]: want to make this a range-based loop?

File: include/aspect/time_stepping/interface.h [hunk 13]
```diff
@@ -308,6 +337,50 @@ namespace aspect
...
+    {
+      AssertThrow(has_matching_plugin<PluginType> (),
+                  ExcMessage("You asked TimeStepping::Manager::get_matching_plugin() for a "
+                             "plugin of type <" + boost::core::demangle(typeid(PluginType).name()) + "> "
+                             "that could not be found in the current model. Activate this "
+                             "plugin in the input file."));
+
+      for (typename std::list<std::unique_ptr<Interface<dim>>>::const_iterator
+           p = active_plugins.begin();
+           p != active_plugins.end(); ++p)
```
Comment (gassmoeller) [inline include/aspect/time_stepping/interface.h:373]: range-based loop?

File: include/aspect/time_stepping/repeat_on_nonlinear_fail.h [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+{
+  namespace TimeStepping
+  {
+    using namespace dealii;
+
+    /**
+     * A class that implements a time stepping plugin to repeat a time step if the
+     * nonlinear solver failed to converge in the specified number of iterations.
+     * This class is automatically enabled if the user specifies that he/she wants
+     * to do this.
```
Comment (gassmoeller) [inline include/aspect/time_stepping/repeat_on_nonlinear_fail.h:37]: Can you add a sentence why the result for the next try would be different from the previous one? Currently it reads as if you may get a different result if you only reran an exactly identical problem.

File: source/simulator/core.cc (hunk 1 above)
Comment (gassmoeller) [reply source/simulator/core.cc:2032]: please check my comment from 2019: https://github.com/geodynamics/aspect/pull/2744/files#r244753861
does this still apply?

File: source/simulator/parameters.cc [hunk 15]
```diff
@@ -265,6 +265,15 @@ namespace aspect
                        "The `Newton Stokes' scheme is deprecated and only allowed for reasons of "
                        "backwards compatibility. It is the same as `iterated Advection and Newton Stokes'.");
 
+    prm.declare_entry ("Nonlinear solver failure strategy", "abort program",
+                       Patterns::Selection("continue with next timestep|cut timestep size|abort program"),
+                       "Select the strategy on what to do if the nonlinear solver scheme fails to "
+                       "converge. The options are:\n"
+                       "`continue with next timestep`: ignore error and continue to the next timestep\n"
+                       "`cut timestep size`: reduce the current timestep size by a factor of 2 and redo "
```
Comment (gassmoeller) [inline source/simulator/parameters.cc:273]: I think it would be better to not hard-code the factor of 2, because it is an input parameter of that plugin. Please update the documentation.

File: include/aspect/parameters.h [hunk 16]
```diff
@@ -97,6 +97,38 @@ namespace aspect
              false;
     }
 
+    /**
+     * A struct that contains the enums to decide what to do when a nonlinear solver fails.
+     */
+    struct NonlinearSolverFailureStrategy
```
Comment (bangerth) [inline include/aspect/parameters.h:103]: This might be more easily done with a namespace. Your `struct` does not have data members.

Comment (bangerth) [reply include/aspect/parameters.h:103]: I recognize that there are other structs of the same kind. I'm going to fix those too right next.

File: source/simulator/solver_schemes.cc [hunk 17]
```diff
@@ -814,6 +814,7 @@ namespace aspect
       }
     while (nonlinear_solver_control.check(nonlinear_iteration, relative_residual) == SolverControl::iterate);
 
+    AssertThrow(nonlinear_solver_control.last_check() != SolverControl::failure, ExcNonlinearSolverNoConvergence());
     signals.post_nonlinear_solver(nonlinear_solver_control);
```
Comment (gassmoeller) [inline source/simulator/solver_schemes.cc:818]: So whether we throw the assert before or after the signal may actually have some impact on the behavior. `post_nonlinear_solver` e.g. informs the postprocessor about the number of nonlinear iterations, and the postprocessor probably keeps accumulating the number of Stokes iterations until it receives the signal that the nonlinear solver has finished. That being said I dont know if the other order would be better, it could mean that one timestep appears twice in the statistics file. Have you check what the output statistics file looks like for a repeated timestep?

File: include/aspect/parameters.h (hunk 16 above)
Comment (bangerth) [reply include/aspect/parameters.h:103]: Ah, drats. These are class members. You can't have a namespace be a class member :-( (That's one of my long-running complaints about C++...)

File: include/aspect/parameters.h [hunk 18]
```diff
@@ -97,6 +97,38 @@ namespace aspect
...
+    {
+      enum Kind
+      {
+        continue_with_next_timestep,
+        cut_timestep_size,
+        abort_program
+      };
+
+      /**
+       * parse the enum value from a string
```
Comment (bangerth) [inline include/aspect/parameters.h:113]:
```suggestion
       * Parse the enum value from a string.
```

File: source/time_stepping/repeat_on_nonlinear_fail.cc [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+    std::pair<Reaction, double>
+    RepeatOnNonlinearFail<dim>::determine_reaction (const TimeStepInfo &/*info*/)
+    {
+      if (nonlinear_solver_just_failed)
+        {
+          nonlinear_solver_just_failed = false;
+          ++current_number_of_repeats;
+
+          if (current_number_of_repeats >= maximum_number_of_repeats)
+            throw "waaa!";
```
Comment (gassmoeller) [inline source/time_stepping/repeat_on_nonlinear_fail.cc:57]: waaa?

File: source/time_stepping/repeat_on_nonlinear_fail.cc [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+namespace aspect
+{
+  namespace TimeStepping
+  {
+    ASPECT_REGISTER_TIME_STEPPING_MODEL(RepeatOnNonlinearFail,
+                                        "repeat on nonlinear solver failure",
+                                        "This time stepping plugin will react when the nonlinear solver "
+                                        "does not converge in the specified maximum number of iterations "
+                                        "and repeats the current timestep with a smaller step size. "
+                                        "This plugin is enabled automatically if \"\" "
```
Comment (gassmoeller) [inline source/time_stepping/repeat_on_nonlinear_fail.cc:122]: parameter name missing in description

File: tests/nonlinear_failure_strategy_cut/statistics [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+# 6: Number of temperature degrees of freedom
+# 7: Number of degrees of freedom for all compositions
+# 8: Number of nonlinear iterations
+# 9: Iterations for temperature solver
+# 10: Iterations for composition solver 1
+# 11: Iterations for Stokes solver
+# 12: Velocity iterations in Stokes preconditioner
+# 13: Schur complement iterations in Stokes preconditioner
+0 0.000000000000e+00 0.000000000000e+00 512 4851 2145 2145 2  0 0  58  60  59 
+1 6.625491997652e+03 6.625491997652e+03 512 4851 2145 2145 8 83 0 338 346 346
```
Comment (gassmoeller) [inline tests/nonlinear_failure_strategy_cut/statistics:15]: So this looks like the statistics file records all nonlinear iterations (2x4) and all Stokes iterations, etc. from all tries. I guess that is fair since we actually made all these iterations?

**Review by gassmoeller — COMMENTED** [review]
Very nice, this will be useful for a bunch of models. Just some smallish comments and a question about the best way to react in the statistics postprocessor.

Also please add a changelog entry.

**Review by bangerth — COMMENTED** [review]
Not getting to the rest of it fast enough, but here's a couple comments.

**Review by bangerth — DISMISSED** [review]
I can't seem to get the uninterrupted time to finish this review, so here's what I have so far. I'll get back to the rest later.

**Comment by tjhei** [comment]
Comments addressed. There are still many tests that need updating, though.

**Comment by tjhei** [comment]
Something is still wrong with nonlinear problems:
```
WARNING: The nonlinear solver in the current timestep failed to converge.
Acting according to the parameter 'Nonlinear solver failure strategy'...
   Postprocessing:
     RMS, max velocity:                               7.49 m/s, 56.3 m/s
     Temperature min/avg/max:                         0 K, 0.5048 K, 1 K
     Heat fluxes through boundary parts:              0 W, 0 W, -3.127 W, 1 W
     Max, min, and rms velocity along boundary parts: 11.29 m/s, 0.001015 m/s, 4.795 m/s, 58.32 m/s, 0.001293 m/s, 23.75 m/s, 30.68 m/s, 0.02087 m/s, 11.42 m/s, 0.04261 m/s, 0.0008854 m/s, 0.03199 m/s

*** Timestep 2:  t=0.0021 seconds, dt=9.27806e-05 seconds
   Solving temperature system... 5 iterations.
   Initial Newton Stokes residual = 3.59864e+06, v = 3.59864e+06, p = 0

   Rebuilding Stokes preconditioner...
   Solving Stokes system... 14+0 iterations.
      Relative nonlinear residual (total Newton system) after nonlinear iteration 1: 1, norm of the rhs: 3.59864e+06

   Solving temperature system... 95 iterations.
[0b3d367e5bc0:34578] *** Process received signal ***
[0b3d367e5bc0:34578] Signal: Floating point exception (8)
[0b3d367e5bc0:34578] Signal code: Floating point divide-by-zero (3)
[0b3d367e5bc0:34578] Failing at address: 0x7fc1f8c5cb14
[0b3d367e5bc0:34578] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x14420)[0x7fc2048bc420]
[0b3d367e5bc0:34578] [ 1] ./libtosi_benchmark_newton_solver.debug.so(_ZNK6aspect13TosiBenchmark12TosiMaterialILi2EE9viscosityEddRKSt6vectorIdSaIdEERKN6dealii15SymmetricTensorILi2ELi2EdEERKNS8_5PointILi2EdEE+0x114)[0x7fc1f8c5cb14]
[0b3d367e5bc0:34578] [ 2] ./libtosi_benchmark_newton_solver.debug.so(_ZNK6aspect13TosiBenchmark12TosiMaterialILi2EE8evaluateERKNS_13MaterialModel19MaterialModelInputsILi2EEERNS3_20MaterialModelOutputsILi2EEE+0x2ff)[0x7fc1f8c5e24f]
[0b3d367e5bc0:34578] [ 3] /__w/aspect/aspect/build/aspect(_ZNK6aspect9SimulatorILi2EE31compute_pressure_scaling_factorEv+0x222)[0x563654f78982]
[0b3d367e5bc0:34578] [ 4] /__w/aspect/aspect/build/aspect(_ZN6aspect9SimulatorILi2EE22assemble_stokes_systemEv+0x101)[0x563655027a71]
[0b3d367e5bc0:34578] [ 5] /__w/aspect/aspect/build/aspect(_ZN6aspect9SimulatorILi2EE36do_one_defect_correction_Stokes_stepERNS_25DefectCorrectionResidualsEb+0x1fa)[0x563654e8b02a]
[0b3d367e5bc0:34578] [ 6] /__w/aspect/aspect/build/aspect(_ZN6aspect9SimulatorILi2EE42solve_iterated_advection_and_newton_stokesEv+0x195)[0x563654e8d835]
[0b3d367e5bc0:34578] [ 7] /__w/aspect/aspect/build/aspect(_ZN6aspect9SimulatorILi2EE14solve_timestepEv+0x1a5)[0x563655a14b25]
[0b3d367e5bc0:34578] [ 8] /__w/aspect/aspect/build/aspect(_ZN6aspect9SimulatorILi2EE3runEv+0x4ef)[0x563655a21ccf]
[0b3d367e5bc0:34578] [ 9] /__w/aspect/aspect/build/aspect(_Z13run_simulatorILi2EEvRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES7_bbbb+0x2cf)[0x5636559ef89f]
[0b3d367e5bc0:34578] [10] /__w/aspect/aspect/build/aspect(main+0x5ad)[0x5636549fb66d]
[0b3d367e5bc0:34578] [11] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3)[0x7fc2046da083]
[0b3d367e5bc0:34578] [12] /__w/aspect/aspect/build/aspect(_start+0x2e)[0x5636549fd11e]
```

**Comment by tjhei** [comment]
@gassmoeller @jdannberg can you remind me what the conclusion of our discussion regarding default of the new setting is?

**Comment by gassmoeller** [comment]
I think we agreed on `continue_with_next_timestep`, because it is closest to the old behavior, and at least provides the user with a warning. I think we also talked about somehow tracking if there are any non-converging timesteps and then issue a warning at the end, because warnings for individual timesteps might get lost in the long output text.

**Comment by naliboff** [comment]
@tjhei - I tested out these changes on top of #4370, and so far using `set Nonlinear solver failure strategy      = cut timestep size` is allowing a complex model to get through non convergent time steps after repeating the time step anywhere from 1-3 times. 

I will report back after additional testing, but so far good news!

**Comment by jdannberg** [comment]
@tjhei From what I remember, this was almost finished, except for the default. Or was there still a problem? 
@naliboff found that this helps with convergence in lithosphere-scale models (and that the results are actually different), so it would be great to get this merged soon!

File: source/simulator/core.cc [hunk 22]
```diff
@@ -1850,91 +1851,125 @@ namespace aspect
     if (parameters.use_operator_splitting)
       compute_reactions ();
 
-    switch (parameters.nonlinear_solver)
+    try
```
Comment (gassmoeller) [inline source/simulator/core.cc:1854]: @naliboff tested this PR for a model, and we just realized that it looks like the mesh deformation above is not reset if the nonlinear solver fails. it would be nice to make the repeat work with mesh deformation as well, but at least we should probably document this. (also I havent tested this extensively yet, maybe it is reset in a place I dont see).

**Comment by tjhei** [comment]
@jdannberg Do you remember what we decided regarding the default? I vaguely remember some kind of statistic to be printed at the end of the run as well? Sorry, I don't remember any of the details.

**Comment by gassmoeller** [comment]
I think we decided to continue on non-convergence, but print a warning (and possibly another warning at the end of the model run if at least one timestep didnt converge). I dont remember about the statistics, maybe it would be useful to report how many timesteps didnt reach the nonlinear solver tolerance?

File: source/simulator/core.cc (hunk 22 above)
Comment (tjhei) [reply source/simulator/core.cc:1893]: I thought this is done here:
https://github.com/geodynamics/aspect/blob/994f686e0d05ea995689c0f19600a11255850e37/source/simulator/core.cc#L2133-L2134

File: tests/global_melt/screen-output [hunk 23]
```diff
@@ -95,3 +101,5 @@ Termination requested by criterion: end time
 
 
 
+
+WARNING: During this computation 2 nonlinear solver failures occurred!
```
Comment (tjhei) [inline tests/global_melt/screen-output:105]: Here is an example of the new output.

**Comment by tjhei** [comment]
Ok, I changed the default to continue and I added printing of information. Any other comments?

File: source/simulator/core.cc (hunk 22 above)
Comment (naliboff) [reply source/simulator/core.cc:1894]: @gassmoeller @tjhei - Yes, the example in question was a version of the continental extension cookbook using #4370 and the commits from this PR. Indeed, in some time steps where the time step was repeated 4-5 times and the time size got quite small (10,000 years -> 500 years) the solver eventually converged, but the solution looked quite different even the nonlinear solve eventually converged. I ran the same model without mesh deformation and the same issue persisted, and I think it is due to the dependence of the rheology on the time step size rather than an issue with this PR. In general, the functionality here helped quite a bit when the final time step size only ends up being a factor of 2 or 4 lower than previous time steps.

**Comment by naliboff** [comment]
> Ok, I changed the default to continue and I added printing of information. Any other comments?

Not for this PR, but based on my testing so far I think it would be good to implement options in the future where:
1. There is a maximum  number of repeated time steps (or minimum time step size) to prevent spurious behavior in models with viscoelasticity
2. An option to repeat the time step and increase the maximum number of nonlinear iterations. This is based on a few cases where the model would have converged with a few more iterations if the maximum number of nonlinear iterations had been set to a higher value. I'll think about ways that this could be automated.

Thanks again for this functionality @tjhei, after a bit of initial testing it also helping in models with two-phase flow and reactions that can lead to rapid changes in porosity.

File: source/simulator/core.cc [hunk 24]
```diff
@@ -1889,91 +1891,127 @@ namespace aspect
...
+                  pcout << "Error: We can not cut the timestep in step 0, so we are aborting."
+                        << std::endl;
+                  throw;
+                }
+              time_stepping_manager.template get_matching_active_plugin<TimeStepping::RepeatOnNonlinearFail<dim>>().nonlinear_solver_has_failed();
+              break;
+            }
+            case Parameters<dim>::NonlinearSolverFailureStrategy::abort_program:
+            {
+              // rethrow the current exception
```
Comment (gassmoeller) [inline source/simulator/core.cc:2008]: the other two options have an additional line that says what is happening next (continue, cut), but this one doesnt, it just throws the error. Can you add another line here, saying that we are now aborting the model run? This would explain to the user we purposefully abort (they would see that from the error message as well, but this makes it more obvious).

File: source/simulator/core.cc [hunk 25]
```diff
@@ -2173,6 +2210,9 @@ namespace aspect
     // throwing an exception. Therefore, we have to do this manually here:
     computing_timer.print_summary ();
 
+    if (nonlinear_solver_failures>0)
```
Comment (gassmoeller) [inline source/simulator/core.cc:2213]:
```suggestion
    if (nonlinear_solver_failures > 0)
```

File: source/time_stepping/repeat_on_nonlinear_fail.cc [hunk 26]
```diff
@@ -0,0 +1,135 @@
...
+
+    template <int dim>
+    void
+    RepeatOnNonlinearFail<dim>::declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Time stepping");
+      {
+        prm.enter_subsection("Repeat on nonlinear solver failure");
+
+        prm.declare_entry("Cut back amount", "0.5",
```
Comment (gassmoeller) [inline source/time_stepping/repeat_on_nonlinear_fail.cc:92]: Maybe and the same for the internal variable name:
```suggestion
        prm.declare_entry("Cut back factor", "0.5",
```

**Review by gassmoeller — COMMENTED** [review]
Thank you! This is almost ready, could you look into my last remaining comments? I think all of @bangerth's comments have been addressed, unless we hear something more I am willing to merge this.

**Comment by tjhei** [comment]
updated.

**Review by gassmoeller — APPROVED** [review]
:+1: Thanks.
